app/models/trace.rb

```diff
@@ -303,9 +305,16 @@ def import
       tp.timestamp = point.timestamp
       tp.gpx_id = id
       tp.trackid = point.segment
-      tp.save!
+      tracepoints << tp
     end
 
+    # Run the before_save and before_create callbacks, and then import them in bulk with activerecord-import
+    tracepoints.each do |tp|
+      tp.run_callbacks(:save) { false }
+      tp.run_callbacks(:create) { false }
+    end
+    Tracepoint.import(tracepoints)
```

Comment:
Shouldn't this be `import!` to raise an exception on error? The old save calls were written that way?

Reply:
Re. batch size: https://github.com/zdennis/activerecord-import#batching mentions some parameter to control the max number of rows per INSERT.

Reply:
I fixed this in 4d45a35f15c7f92c125ff070bf4757adc064ef9f

Reply:
Well that stops us generating one mammoth SQL statement yes, but it still creates in-memory objects for all the points. Not sure whether that is actually an issue or not?

Reply:
Good point, but I don't know, tbh.

I can look at doing something clever to yield the points in batches too.

Reply:
OK, I've found a way to use `enum_for` to allow a simple way to pull points out of gpx.points using `.each_slice`. That means we will do multiple bulk_imports but without having more than 1,000 points in memory at one time.